### PR TITLE
Fix panic in gRPC UnaryServerInfo

### LINKED ISSUE #691: grpctrace addEventForMessageSent tries to type-assert nil to proto.Message
The unary server interceptor for gRPC panics when I return an error from my handler:

```go
return nil, status.Errorf(codes.PermissionDenied, "Permission denied")
```

My suggested fix is to add a "message" event for the status proto if `err != nil` instead of adding an event for `resp` (which is `nil`).

```diff
diff --git a/plugin/grpctrace/interceptor.go b/plugin/grpctrace/interceptor.go
index 344dccf..0f4363e 100644
--- a/plugin/grpctrace/interceptor.go
+++ b/plugin/grpctrace/interceptor.go
@@ -302,11 +302,12 @@ func UnaryServerInterceptor(tracer trace.Tracer) grpc.UnaryServerInterceptor {
 
 		resp, err := handler(ctx, req)
 
-		addEventForMessageSent(ctx, 1, resp)
-
 		if err != nil {
 			s, _ := status.FromError(err)
 			span.SetStatus(s.Code(), s.Message())
+			addEventForMessageSent(ctx, 1, s.Proto())
+		} else {
+			addEventForMessageSent(ctx, 1, resp)
 		}
 
 		return resp, err
```

I can submit a pull request later.

@mujtaba-ahmed12 maybe good to add a test case for this too when you work on #684.

### PR DESCRIPTION
**Pull request by MrAlias**
Fixes unresolved issue identified in #691 and attempted in #697.

Adds unit test to ensure the `UnaryServerInfo` function does not panic during an error returned from the handler and appropriately annotates the span with the correct event.

Restructures the interceptor to remove this class of errors.

Actually Fixes #691

### CODE REVIEW

File: plugin/grpctrace/interceptor.go [hunk 1]
```diff
@@ -43,9 +43,29 @@ var (
...
+	span := trace.SpanFromContext(ctx)
+	if p, ok := message.(proto.Message); ok {
+		span.AddEvent(ctx, "message",
+			messageTypeKey.String(string(m)),
+			messageIDKey.Int(id),
+			messageUncompressedSizeKey.Int(proto.Size(p)),
+		)
+	} else {
+		span.AddEvent(ctx, "message",
+			messageTypeKey.String(string(m)),
```
Comment (lizthegrey) [inline plugin/grpctrace/interceptor.go:60]: should this be string() or should it utilize Sprint()?

Comment (MrAlias) [reply plugin/grpctrace/interceptor.go:60]: `string()` is correct here as it is converting to the underlying type of the `messageType` (`string`). There is no `String()` method.

Comment (lizthegrey) [reply plugin/grpctrace/interceptor.go:60]: still confused -- is m a generic `interface{}` or is it just a handy way to pass either a `Proto` or a `string`?

Comment (lizthegrey) [reply plugin/grpctrace/interceptor.go:60]: also, please consider a type switch here, as it makes the behavior a little bit more obvious.

Comment (MrAlias) [reply plugin/grpctrace/interceptor.go:60]: Hmm, I think there is confusion between `m` with `message`. `m` is a `messageType`, a new type added to ensure only two message types are allowed.

Comment (lizthegrey) [reply plugin/grpctrace/interceptor.go:60]: oh, aha, yup, that would do it. having one variable named `m` and the other named `message` is incredibly confusing.

File: plugin/grpctrace/interceptor.go [hunk 2]
```diff
@@ -43,9 +43,29 @@ var (
 	messageUncompressedSizeKey = kv.Key("message.uncompressed_size")
```
Comment (vmihailenco) [inline plugin/grpctrace/interceptor.go:43]: Should this be `messaging.message_payload_size_bytes` according to https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md or these are unrelated?

**Pull request merged**